### DISCUSSION TITLE
Disable tf4micro agc and ns training for mWW

### DISCUSSION THREAD

**Discussion by StuartIanNaylor** [FutureProofHomes/Satellite1-Hardware#9] [Ideas]
The specifically purchased Xmos for speech enhancement has a more advanced tflite model based NS Voice extraction scheme, but the pipeline still forces audio through the NS and AGC of TF4micro.
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/experimental/microfrontend
Even if of comparitive quality to the Xmos solution its a duplication of ops that is done 1st on the Xmos.

https://github.com/esphome/home-assistant-voice-pe/issues/279#issuecomment-2598495668

The reply of 'no plans' uses a strange chicken and egg argument, whilst the Xmos audio is there being forced through TF4micro NS and AGC which is trained in and it would seem no-one has tried to train with those features disabled to try the resultant model with the frontend features turned off.

Capturing the audio from the xmos algs via firmware to provide a constant wyoming broadcast or enable the USB Audio class driver provided by Xmos should be fairly straight forward.
The professional way would be to run the xmos tflite and process algs on a dev machine and force a dataset through processing at far higher rates than sample rate limits of the mic hardware.
Likely a constant Wyoming broadcast would be the easiest method as it just means removing mWW from the already implemented pipeline.

This leads into the training as the Xmos has been purchased in to provide 3m farfield voice extraction and due to being blind to what the Xmos algs do RiRs recorded @ 1.5m from a single mic are being applied to create a training dataset.
https://github.com/kahrendt/microWakeWord/issues/28#issuecomment-2564400870
I don't think its possible not to be critical of the current dataset creation from Piper creating a 1000 samples with little variation to applying pre recorded RIRs @ a single fixed distance of many locations such as forrests and massive structures for a device that is to sit at home and provide 3m farfield in relatively normal sized rooms.
Where supposedly because the xmos alg is a farfield voice extraction these Rirs should already attentuated but unfortunately devs and community are currently totally blind to what level.
What RIRs and why so much research and papers at the start os the audio pipeline to deverberation and clean audio before processing is due to the huge problem you can not just train it in to a basic classification alg of a mixed resultant mono audio stream.
https://github.com/OHF-Voice/wake-word-collective/issues/11#issuecomment-2543304585

There are so many basic 101 errors to the whole approach to the hardware purchased in and model training that if someone comfortable with the Microcontroller skills could provide the output from the Xmos algs then at least the community would not be blind.

I have tacked on some ideas to https://github.com/kahrendt/microWakeWord/issues/28#issuecomment-2564400870 but this blind situation is a crazy situation that surely can be rectified.
The missconceptions about training are huge especially the importance of the size, quality of needed datasets.
https://developers.google.com/machine-learning/crash-course/overfitting/data-characteristics

Even if tf4micro doesn't work with a ondevice training scheme its so shortsighted not to be capturing actual use data, that maybe the recouped ops of NS & AGC could create a rolling window to capture the KW aswell as streamed voice command...
https://ai.google.dev/edge/litert/models/ondevice_training
That data with an opt-in could be submitted to create over time gold standard datasets so opensource on a equal footing to big data.
Why current approaches seem so out of line with basic ML concepts where we have no idea to the data we are recieving and the current dataset is created on total guesswork...

Sadly when you do point these basic errors out they are being closed off without even approaching the problems in what can be said as closed source methods and actions.

**Comment by that1guy** [comment]
I'm going to tap in @gnumpi to provide his thoughts here.  

@gnumpi?

**Reply by StuartIanNaylor** [reply]
One way you could improve the dataset problem is to use https://ai.google.dev/edge/litert/models/ondevice_training where a small captured dataset biases the weights of a larger pretrained model.
Dunno if you can do that with tf4micro but with tflite or LiteRT as its now called you can train in environment of use and its users to make it more accurate.

Thing is to start capturing data especially ondevice with good metadata to an opensource repo via an opt-in as the Google ML Crash Course hints at how far short current training is in dataset size.

**Reply by StuartIanNaylor** [reply]
Also the Mobilenet model chosen for mWW is much larger in terms of paramters https://arxiv.org/pdf/1907.09595 as quoted as the model in https://github.com/kahrendt/microWakeWord?tab=readme-ov-file whilst https://github.com/google-research/google-research/blob/master/kws_streaming/experiments/kws_experiments_12_labels.md#bc_resnet_1 can have as low as 10k params and is also supposedly streamable.

**Reply by StuartIanNaylor** [reply]
I had a go with https://github.com/google-research/google-research/tree/master/kws_streaming and a CRNN I know with synthetic data of my own creation, no RIR as how much no-one knows.
The resultant tflite model and test script is in https://github.com/StuartIanNaylor/tf-kws, augment scripts in https://github.com/StuartIanNaylor/create-dataset, dataset selection in https://github.com/StuartIanNaylor/syn-dataset
Just to test 'My Way' as best I can with synthetic and still far too few voices, on-device and real much more preferable but a bigger dataset of more voice variance would likely be best.
Also the dataset was recorded at a too high volume without enough headroom but actually did training on CPU and it will do as an example that can still be improved.
I find using the 'Noise' label is an inverse VAD and no VAD model is really needed.
I have the actual dataset here if someone wants to try training it and compare the tflite results of mWW...
 https://drive.google.com/file/d/1QyzbPWyls253lLBb3Wn-vlg1T_qMHguJ/view?usp=drive_link
 tflite with flex delegate here https://drive.google.com/file/d/1eHcoMFNgbFUlro3T28jSipAIaAsI0965/view?usp=drive_link

**Comment by gnumpi** [comment]
Sounds like a valid point you came up with. I have to say that I haven’t spent much time into improving mWW yet. It’s on my todo list, if it is ok I will ping you when I get there and we can further discuss this. Thanks for reporting your findings here!

**Reply by StuartIanNaylor** [reply]
Yeah of course, ping anytime.
I deliberately did a different model CRNN-State from the Goggle KWS-streaming repo purely as a Tflite model comparison.
The dataset they create is without doubt hugely overfitted and the RIR bit is just strange.
I am in no rush as looking for a TTS to provide more 'voices' emotivoice has a lot of voices but some sound a bit synthetic.
https://github.com/coqui-ai/TTS  ⓍTTSv2 does a great job and maybe cloning further voices is a slowe but better idea.
Capturing actual device recordings is and will always be best however.
I don't know if the training in https://github.com/kahrendt/microWakeWord is inline with current mWW code...
I will try to hack out the dataset creation and just train a mWW tflite model but its a pain as the dataset is stored as ready made spectrogram, which is another thing I don't like as listening to wavs is easy but in spectrogram its not really possible.
The situation with a fairly poor dataset likely contributes to results more than the tf4micro frontend AGC + NS but will try to create a model of one of 'my' datasets with them enabled and one without.
Being blind without samples of KW and command sentences at various distances of varying volumes means I haven't a clue what sort of scope the current arrangement of AGC provides.
The dataset should resemble average normal use, but just don't know what that is in terms of samples.

**Reply by StuartIanNaylor** [reply]
> Thanks for reporting your findings here!

I have reported on HA in various threads, on the actual repo's and obviously here and been waiting patiently but seems like no reply is the answer from Kevin or anyone at HA.
I am not sure whats happening added a newer tflite to https://github.com/StuartIanNaylor/tf-kws with the last 3 latest repos all being dataset and kws scripts getting my memory back to experiments I did 3 years ago.
I did another dataset bigger with more voices and no RIRs recorded in a forest!
https://drive.google.com/file/d/1vodwYOIagRdJXJ2yCcmX7kHlWbVmHZb0/view?usp=drive_link
Its grown in size but nothing near what Google advises https://developers.google.com/machine-learning/crash-course/overfitting/data-characteristics

**Reply by StuartIanNaylor** [reply]
PS I have been keeping quiet just to see if the devs do have any idea on how datasets should be creared, which I will tell you guys here as really it should be obvious.
The Xmos uses a tensorflow model and AGC for farfield, the AGC should be fairly easy to port to X86 and the platform agnostic TFlite is already available.
Using something like GPUrir or PyroomAcoustics you apply a different RIR to the noise sample and foreground samples such the KW. Generally that is how noise and conflicting audio is and the point sources are often different and you record multi-channel audio where the recievers of GPUrir or PyroomAcoustics are setup as the mics of hardware (x2 or x4)
You run the dataset and augment with the algs of use and your dataset should be exactly the result of what the hardware produce and not this strange guesswork or feeding it with anything they can find.

Thats it and also this should also extend to ASR as will likely create more accurate results as Big Data creates finely tuned systems and if you want to come close you will have to start training as a system where the devices of use will create artifacts and have a signature that should be in the dataset.

We will see how things go as maybe you might not hear from me again as I think its great Speech-to-Phrase turned up, by why 3 years, why does it seem it always has to be refactored and ownbrand code to the exclusion of existing and that it should also be a skill router.

Dunno what is happening but microWakeWord and any dataset can make models more accurate by actually having data of use and often you can quite accurately recreate synthetic data by running through the actual processes on a heavy weight workstation or server to create datasets at much faster than realtime. Real capture is better but it needs accurate and detailed metadata so you can make sure your not overfitting to any traits it might contain.

**Comment by StuartIanNaylor** [comment]
PS I thought as much as no longer able to comment on the forums and if you read as usual the HA reason is total BS as all I have posted is solutions and ideas and posted where things are wrong but "Account suspended: No constructive purpose to their actions other than creating dissent within the community" is what you supposedly are doing.

This all goes back to 22 where I posted some ideas https://community.rhasspy.org/t/thoughts-for-the-future-with-homeassistant-rhasspy/4055/3
Various posts on why Whisper to later ones advocating once more Wenet https://community.rhasspy.org/t/state-of-the-art-as-it-were/4973/12

Where to 
https://community.rhasspy.org/t/feel-a-little-like-i-am-in-limbo-now-ha-vs-rhasspy/4768/6?u=rolyan_trauts
I recieved this email and once more what is said is totally untrue as never was shit said, just why certain methods are inferior and why other methods are better, this time it was esp32.

> Paulus Schoutsen<paulus@paulusschoutsen.nl>
> ​
> You
> ​
> ​
> Michael Hansen
> ​
> Hey Stuart,
> 
> Adding Mike to the CC.
> 
> I've been reading along with your posts for a while now both in the Rhasspy and the Home Assistant community forums. Although you might intend well, your posts are derailing threads and you need to stop posting. Mike agrees with this.
> 
> We are not interested in hearing that everything we did is crap, everything we are working on is crap, or hear about solutions that are unrealistic and unachievable.
> 
> Consider this a courtesy warning that these posts need to stop.
> 
> Paulus

You can not give critique as just like closed source it will be censored. Even worse as I am not bothered that my suggestions ingnored for 3 years have finnally started to take shape with Speech2Phraise the blantant refactoring and rebranding of the Wenet system of ngrams is just really bad plaguarism and at least I am honest as opposed to certain others.

Shithausen and synesthesiam make good partners and as I have been banned maybe someone else will question if this copying, refactoring and rebranding to ignore great existing opensource to partition into smaller herds whilst causing huge delays in adoption by certain devs only providing own code of what already exists but refactored and rebranded.
Its not the 1st and I have kept quiet about this for some time but it probably needs some discussion as also the absolute BS mechanism of creating your own OpenHomeFoundation so you can claim proprietory tech is actually open standards...

Anyway personally I am quite happy as there are certain things in opensource I find distasteful and from Mycroft, Rhasppy to HA VA I am not sure why but its has attracted a certain type and a hell of a lot of snakeoil. Much current I am glad to no-longer be part of.

I am doing a little bit more research on classification KWS and how you can make them more accurate and will post that info in another thread when completed.
I am embarised about advocating Wenet ngram domain specific language models so much to what has happened as they at least deserve a mention.
https://wenet.org.cn/wenet/lm.html

I think really so you can keep a small domain specific ngram language model you should have a front skill router. That then routes the ASR AUDIO to ASR Skill servers. Entity lists can grow large and no longer remain small or domain specific just as say a music library of album names, tracks and bands. So having a predicate skill router you can make multi-modal domains that are partitioned to a skill domain and at least keep that domain as small as possible.
Also when no skill predicate is found the likes of Whisper and a LLM can still be implemented as a catch all.

The only thing about Speech2Phraise is ignoring great NLP libaries such as Spacy or NLTLK as they are likely too complex and would take too much time and be very obvious if just refactored and rebranded but there use would likely create a much better system.

Do as you wish its your choice and if you agree or not its an open space and its all just opinion, but I am quite happy to be excluded from some extremely dubious opensource.

**Reply by that1guy** [reply]
Hi Stuart, thanks for all the thoughts here.  Myself and the small team are focused on important milestones.  Sorry for the delayed response.  Feel free and open this back up and we'll follow up just as soon as we can!

**Reply by StuartIanNaylor** [reply]
You have the basis and the links to the basic scripts repos I did for creating the dataset.
Basically its have far more voice variance, use real not synthetic, but we use synthetic as no real avail.
Give the classification model something to train against, create a falseKW that contains soundlikes of similar intonation length.
We might not have the millions to billion dataset items of Google recomendation, but have far more than is currently prescribed.
Create multichannel wavs of noise and voice mixtures via gpuRir with  https://github.com/yluo42/TAC/blob/master/data/README.md showing a good gpuRIR implementation so that can be run through the actual algs of use from the xMOS.
The dataset should be what is the input feed to the model...

Always give credit where its due and don't leach, refactor and rebrand :)

Its when you can and if you have to as really this has been freely posted on the HA VA repos as they are killing microWakeWord with a terrible training implementation that makes zero sense.
You shouldn't really have to, but might have to due to this crazy closed method they operate in, if delay in updates goes on too long.

**Reply by StuartIanNaylor** [reply]
@that1guy I did a bit more training and happy with what I got as the noise index can be used as a inverse VAD so no extra VAD model needed.
https://github.com/StuartIanNaylor/tf-kws/blob/main/wake_word_benchmark-stream.py https://github.com/StuartIanNaylor/tf-kws/blob/main/wake_word_benchmark-non-stream.py can be used to test the tflite models in that repo such as https://github.com/StuartIanNaylor/tf-kws/blob/main/stream_state_external.tflite the 2prefix is the original with only 4 labels such as https://github.com/StuartIanNaylor/tf-kws/blob/main/2stream_state_external.tflite and likely the noise index isn't too great for inverse VAD but the introduction of another label PhoneKW makes the inverse VAD and KW spotting more accurate.

Doesn't matter that its the CRNN from streamingKWS as it just proves its the dataset not the KWS that is of importance.
Like the porcupine benchmark the Demand dataset can be downloaded from https://www.kaggle.com/aanhari/demand-dataset and the LibriSpeech here http://www.openslr.org/resources/12/test-clean.tar.gz .
They have a Jarvis but not a 'Hey Jarvis' and really should have external recordings but this will do for a test.
https://drive.google.com/file/d/1Q8KFDm61XLONu3D40BP1B3Y4cToQZLFR/view?usp=drive_link

The accuracy greatly exceeds the picovoice benchmark and would be interesting to see how porcupine does in my version.
I looked at there benchmark code and don't really care about pocketsphinx or snowboy as Porcupine isn't that great for accuracy but they are terrible and a single hacked script is all that is needed IMO.

**Reply by StuartIanNaylor** [reply]
I should of said non stream are in the same repo https://github.com/StuartIanNaylor/tf-kws but here a gdrive links also
https://drive.google.com/file/d/1p7mG4MB2ipWrL5FV-v-FkhX5n_Nx8eG4/view?usp=drive_link
https://drive.google.com/file/d/1Zsf_qEsKY5Ry0LZtBVoG4tAkrj2wfncR/view?usp=drive_link

**Reply by StuartIanNaylor** [reply]
@that1guy PS they have done it again.
[r/homeassistant]You have been muted from r/homeassistant
You will not be able to message the moderators of [r/homeassistant](https://www.reddit.com/r/homeassistant) for 28 days.
Its really bad as this is so against everything in opensource and its just fact that has been posted.

https://www.reddit.com/r/homeassistant/comments/1kf79it/comment/mqrvf0l/?context=3

There are people in the community beginning to realise...
